### PR TITLE
Adds full swagger-based docs

### PR DESCRIPTION
**Pull request by carsonfarmer**
No major code changes here, just tons of docs! There are bound to be copy-pasta mistakes in here... so feel free to nit, or pick some of your fav to double check. Then we can just create tickets for issues as they crop up.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -22,6 +22,7 @@ lint:
 	golint `go list ./... | grep -v /vendor/`
 
 build:
+	swag init -g core/api.go
```
Comment (carsonfarmer) [inline Makefile]: Now it'll automatically rebuild docs, which doesn't take a ton of extra time, but we _might_ want to think about having this in a separate make target?

File: core/api_contacts.go [hunk 2]
```diff
@@ -23,6 +36,10 @@ func (a *api) addContacts(g *gin.Context) {
 		g.String(http.StatusBadRequest, "invalid contact")
 		return
 	}
+	if contact.Id != g.Param("id") {
```
Comment (carsonfarmer) [inline core/api_contacts.go]: as per separate conversation

File: core/api_invites.go [hunk 3]
```diff
@@ -10,6 +10,19 @@ import (
 	"github.com/mr-tron/base58/base58"
 )
 
+type InviteInfo map[string]string
```
Comment (carsonfarmer) [inline core/api_invites.go]: annoying quirk of swaggo... doesn't allow you to have 'generic' return params

File: core/api_logs.go [hunk 4]
```diff
@@ -9,6 +9,20 @@ import (
 	"github.com/gin-gonic/gin"
 )
 
+type SubsystemInfo map[string]string
```
Comment (carsonfarmer) [inline core/api_logs.go]: same quirk again

File: core/api_threads.go [hunk 5]
```diff
@@ -114,6 +135,11 @@ func (a *api) addOrUpdateThreads(g *gin.Context) {
 		return
 	}
 
+	if thrd.Id != g.Param("id") {
```
Comment (carsonfarmer) [inline core/api_threads.go]: as per separate convo

**Comment by carsonfarmer** [comment]
When reviewing this one @sanderpick, let me know if you think we should make serving the local `docs` endpoint optional? Additionally, consider the docs build step, and let me know if we should move that to its own call and step?

File: Makefile (hunk 1 above)
Comment (sanderpick) [reply Makefile]: seems like a good idea to have it separate, build is for the binary

**Comment by sanderpick** [comment]
Damn, lots of docs in there! I'm a little concerned w/ having to maintain the descriptions between the the REST API, go client, and all future clients... thoughts?

**Comment by sanderpick** [comment]
Personally, I would keep the descriptions in the REST API dead simple, if any at all... and leave the detailed descriptions to the client tooling, which is really where folks should be poking around. I'm not suggesting you delete stuff now, I just didn't realize your approach here.

**Comment by sanderpick** [comment]
For the docs endpoint, possible to make it a flag on daemon start, like `--swagger` or something?

**Comment by carsonfarmer** [comment]
Re: too much description, yes this is a valid concern. Didn't occur to me until I was deep into it, but valid point. We can actually mostly 'drop' the 'description' option and stick with just 'summary'.

**Comment by carsonfarmer** [comment]
Re: `--swagger` flag, yup, easy to add this. Could be nice. What makes more sense, `--docs` or `--swagger` (just wondering if swagger is less informative than docs).

**Comment by sanderpick** [comment]
Ah yeah, `--docs` seems better

**Comment by carsonfarmer** [comment]
Used `--serve-docs` because `daemon` already has a short code `-d` for `--debug` and didn't want to mess with that. Also moved the docs build code to its own `build_docs` target in `Makefile`.

**Comment by carsonfarmer** [comment]
Suck in some extra fixes there for ya @sanderpick. Should be GTG now.

File: Makefile (hunk 1 above)
Comment (carsonfarmer) [reply Makefile]: Agreed, moved

**Pull request merged**